File: _tech/algo/dynamic/robbery/linear/rob.go
func max(a int,b int) int{
    if a>b{
        return a
    }
    return b 
}

func rob(nums []int) int {
    if len(nums) == 0 {
        return 0
    }
    p1, p2 := 0, 0 
    for _, num := range(nums) {
        p1 , p2 = max(p2+num, p1), p1
    } 
    
    return p1 
}

File: _tech/algo/dynamic/robbery/linear/rob.py
def rob(nums: List[int]) -> int:
    if not nums:
        return 0

    dp = {}
    n = len(nums)

    def decide(i):
        if i < 2:
            return max(nums[: i + 1])

        if i not in dp:
            rob = decide(i - 2) + nums[i]
            skip = decide(i - 1)
            dp[i] = max(rob, skip)
            
        return dp[i]


    return decide(n - 1)
